private/extensions/maven.bzl

```diff
@@ -0,0 +1,80 @@
+load("@bazel_skylib//lib:sets.bzl", "sets")
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_file")
+load("//:coursier.bzl", "pinned_coursier_fetch")
+load("//:specs.bzl", _json = "json", "parse")
+load(":download_deps.bzl", "download_deps")
+
+_install = tag_class(
+    attrs = {
+        "name": attr.string(
+            default = "maven",
+        ),
+        "artifacts": attr.string_list(
+            default = [],
+        ),
+        "lock_file": attr.label(),
+        "repositories": attr.string_list(
+            default = ["https://repo1.maven.org/maven2"],
+        ),
+    }
+)
+
+def _maven_impl(mctx):
+    # Gather the artifacts
+    repos = {}
+
+    for mod in mctx.modules:
```

Comment:
This is exciting :) have you tried making this work in a transitive project environment?